request.go

```diff
@@ -140,6 +140,91 @@ type Request struct {
 	cancelCtx context.CancelFunc
 }
 
+// This is essentially a more unmarshaled Request.
+// I say more unmarshaled, because StatAttrs is a FileStat, which while better
+// than a binary array, is still supposed to be sftp protocol values instead of
+// something like os or fs package types and values.
+type ParsedRequest struct {
+	Method    string
+	Filepath  string
+	OpenFlags FileOpenFlags
+	StatFlags FileAttrFlags
+	StatAttrs FileStat
+	Target    string // for renames and sym-links
+}
+
+// Marshal is used to turn a ParsedRequest into a Request.
+func (pr *ParsedRequest) Marshal() (r *Request) {
+	r = NewRequest(pr.Method, pr.Filepath)
+
+	switch pr.Method {
+	case "Get", "Put":
+		r.Flags = 0
+		if pr.OpenFlags.Read {
+			r.Flags |= sshFxfRead
+		}
+		if pr.OpenFlags.Write {
+			r.Flags |= sshFxfWrite
+		}
+		if pr.OpenFlags.Append {
+			r.Flags |= sshFxfAppend
+		}
+		if pr.OpenFlags.Creat {
+			r.Flags |= sshFxfCreat
+		}
+		if pr.OpenFlags.Trunc {
+			r.Flags |= sshFxfTrunc
+		}
+		if pr.OpenFlags.Excl {
+			r.Flags |= sshFxfExcl
+		}
+
+	case "Stat", "Setstat":
+		var buf []byte
+		r.Flags = 0
+
+		if pr.StatFlags.Size {
+			r.Flags |= sshFileXferAttrSize
+			buf = marshalUint64(buf, pr.StatAttrs.Size)
+		}
+		if pr.StatFlags.UidGid {
+			r.Flags |= sshFileXferAttrUIDGID
+			buf = marshalUint32(buf, pr.StatAttrs.UID)
+			buf = marshalUint32(buf, pr.StatAttrs.GID)
+		}
+		if pr.StatFlags.Permissions {
+			r.Flags |= sshFileXferAttrPermissions
+			buf = marshalUint32(buf, FromFileMode(os.FileMode(pr.StatAttrs.Mode)))
+		}
+
+		if pr.StatFlags.Acmodtime {
+			r.Flags |= sshFileXferAttrACmodTime
+			buf = marshalUint32(buf, pr.StatAttrs.Atime)
+			buf = marshalUint32(buf, pr.StatAttrs.Mtime)
+		}
```

Comment:
There is a very subtle difference with the new logic, but not one that I expect to be significant:

There is no FileAttrFlags.ExtendedData, so I simply look for there being data in fileStat.Extended.

Otherwise this is now combined, and the logic should be unchanged.

Especially since fileStatFromInfo won't set fileStat.Extended unless len(fileStat.Extended) > 0, and since that's the exact same check I'm doing...